libs/core/include/core/synchronisation/protected.hpp

```diff
@@ -43,58 +43,26 @@ class Protected
   Protected &operator=(Protected &&) = delete;
 
   template <typename Handler>
-  void ApplyVoid(Handler &&handler)
+  constexpr decltype(auto) Apply(Handler &&handler)
```

Comment:
`decltype(auto)` is not appropriate here. Use `auto` instead to avoid leaking references to the protected payload.